### PR TITLE
Draft: [luci/service] Support StridedSlice dynamic shape inference

### PR DESCRIPTION
**Pull request by qsunki**
This supports dynamic shape inference for StridedSlice.
    
ONE-DCO-1.0-Signed-off-by: sunki <ajsthfldu@outlook.com>

---

for: https://github.com/Samsung/ONE/issues/13697

### CODE REVIEW

**Comment by qsunki** [comment]
Typo In first commit 😂
> This adds test cases about StrideSlice's Shape Inference.
I will fix.

File: compiler/luci/service/src/Nodes/CircleStridedSlice.test.cpp [hunk 1]
```diff
@@ -41,3 +42,351 @@ TEST(CloneNodeTest, clone_StridedSlice)
...
+  luci::CircleConst strides;
+
+  luci::CircleStridedSlice strided_slice;
+
+  input.shape({5, 6, 7, 8});
+  input.shape_status(luci::ShapeStatus::VALID);
+
+  begin.dtype(loco::DataType::S32);
+  begin.shape({4});
+  begin.size<loco::DataType::S32>(4);
```
Comment (pcs1265) [inline compiler/luci/service/src/Nodes/CircleStridedSlice.test.cpp:60]: Do we need to set the size of tensor in CircleConst twice?
As far as I know, when we call the `shape()` or `size()` function, it eventually resizes the vector that it contains.

File: compiler/luci/service/src/Nodes/CircleStridedSlice.test.cpp [hunk 2]
```diff
@@ -41,3 +42,351 @@ TEST(CloneNodeTest, clone_StridedSlice)
...
+  ASSERT_TRUE(shape.dim(3).known());
+  ASSERT_TRUE(shape.dim(4).known());
+  ASSERT_EQ(1, shape.dim(0).value());
+  ASSERT_EQ(0, shape.dim(1).value());
+  ASSERT_EQ(5, shape.dim(2).value());
+  ASSERT_EQ(7, shape.dim(3).value());
+  ASSERT_EQ(8, shape.dim(4).value());
+}
+
+TEST(ShapeRuleTest, strided_slice_NEG_1)
```
Comment (pcs1265) [inline compiler/luci/service/src/Nodes/CircleStridedSlice.test.cpp:228]: We could give more specific names to the tests, I think.

File: compiler/luci/service/src/ShapeInfer_StridedSlice.h [hunk 3]
```diff
@@ -23,9 +23,13 @@
 
 namespace luci
 {
+namespace sinf
+{
 
 loco::TensorShape infer_output_shape(const CircleStridedSlice *node);
```
Comment (pcs1265) [inline compiler/luci/service/src/ShapeInfer_StridedSlice.h:29]: Just wondering, was this function declared here before you wrote the code for this PR?

File: compiler/luci/service/src/Nodes/CircleStridedSlice.test.cpp (hunk 2 above)
Comment (qsunki) [reply compiler/luci/service/src/Nodes/CircleStridedSlice.test.cpp:228]: Thank you, I will rename it a more specific.

File: compiler/luci/service/src/Nodes/CircleStridedSlice.test.cpp (hunk 1 above)
Comment (qsunki) [reply compiler/luci/service/src/Nodes/CircleStridedSlice.test.cpp:60]: As far as I know, shape manipulates rank and dim, while size manipulates the data of `CircleConst`.

File: compiler/luci/service/src/Nodes/CircleStridedSlice.cpp [hunk 4]
```diff
@@ -33,4 +35,14 @@ luci::CircleNode *CloneNodeLet<CN::STUV>::visit(const luci::CircleStridedSlice *
   return cloned;
 }
 
+namespace sinf
+{
+
+loco::TensorShape Algorithm::visit(const luci::CircleStridedSlice *node)
+{
+  return infer_output_shape(node);
```
Comment (llFreetimell) [inline compiler/luci/service/src/Nodes/CircleStridedSlice.cpp:43]: Please move the codes of `infer_output_shape` in this file :)
i.e.,

```suggestion
  // implementation codes of infer_output_shape function
```

Comment (qsunki) [reply compiler/luci/service/src/Nodes/CircleStridedSlice.cpp:43]: Thank you for the feedback! I'll move the `infer_output_shape` codes and remove the mentioned files.

File: compiler/luci/service/src/Nodes/CircleStridedSlice.cpp [hunk 5]
```diff
@@ -33,4 +46,418 @@ luci::CircleNode *CloneNodeLet<CN::STUV>::visit(const luci::CircleStridedSlice *
...
+  for (uint32_t idx = 0; idx < shape_size; ++idx)
+  {
+    int64_t dim = output_shape_vector.at(shape_size - 1u - idx);
+    bool known = output_known_vector[shape_size - 1u - idx];
+    LUCI_ASSERT(0 <= dim && dim < 0xfffffffL, "Dimension size exceeds limit");
+    // reverse copy
+    if (known)
+    {
+      output_shape.dim(idx) = static_cast<uint32_t>(dim);
+    }
```
Comment (icodo98) [inline compiler/luci/service/src/Nodes/CircleStridedSlice.cpp:455]: (Optional)
```suggestion
    bool known = output_known_vector[shape_size - 1u - idx];
    if(not known) 
      continue;
    int64_t dim = output_shape_vector.at(shape_size - 1u - idx);
    LUCI_ASSERT(0 <= dim && dim < 0xfffffffL, "Dimension size exceeds limit");
    // reverse copy
    output_shape.dim(idx) = static_cast<uint32_t>(dim);
```

Comment (qsunki) [reply compiler/luci/service/src/Nodes/CircleStridedSlice.cpp:455]: good!

**Pull request closed without merging**